contributors/design-proposals/all-in-one-volume.md

```diff
@@ -0,0 +1,297 @@
+### ConfigMap + Secrets
+
+Again, the user wishes to deploy containers involving configuration data and
+passwords. This time the user is executing an Ansible playbook stored as a
+configmap, with some sensitive encrypted tasks that are decrypted using a vault
+password file.
+
+### ConfigMap + Downward API
+
+In this case, the user wishes to generate a config including the pod’s name
+(available via the metadata.name selector). This application may then pass the
+pod name along with requests in order to easily determine the source without
+using IP tracking.
+
+### Secrets + Downward API
+
+A user may wish to use a secret as a public key to encrypt the namespace of
+the pod (available via the metadata.namespace selector). This example may be
+the most contrived, but perhaps the operator wishes to use the application to
+deliver the namespace information securely without using an encrypted
+transport.
+
+### Collisions between keys when configured paths are identical
+
+In the event the user specifies any keys with the same path, the pod spec will
+not be accepted as valid. Note the specified path for mysecret and myconfigmap
+are the same:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: volume-test
+spec:
+  containers:
+  - name: container-test
+    image: busybox
+    volumeMounts:
+    - name: all-in-one
+      mountPath: "/system-volume"
+      readOnly: true
+  volumes:
+  - name: all-in-one
+    system:
+      sources:
+      - secret:
+          name: mysecret
+          items:
+            - key: username
+              path: my-group/data
+      - configMap:
+          name: myconfigmap
+          items:
+            - key: config
+              path: my-group/data
+```
+
+
+### Collisions between keys without configured paths
+
+The only run time validation can occur is when all the paths are known at pod
+creation, similar to the above scenario. Otherwise, when a conflict occurs the
+most recent specified resource will overwrite anything preceding it (this is
+true for resources that are updated after pod creation as well).
+
+### Collisions when one path is explicit and the other is automatically projected
+
+In the event that there is a collision due to a user specified path matching
+data that is automatically projected, the latter resource will overwrite
+anything preceding it as before.
+
+## Code changes
+
+### Proposed API objects
+
+```go
+type SystemProjections struct {
+    Sources           []SystemVolumeProjection `json:"sources"`
+    DefaultMode       *int32                   `json:"defaultMode,omitempty"`
+}
+
+type SystemVolumeProjection struct {
+    Secret      *SecretVolumeSource      `json:"secret,omitempty"`
+    ConfigMap   *ConfigMapVolumeSource   `json:"configMap,omitempty"`
+    DownwardAPI *DownwardAPIVolumeSource `json:"downwardAPI,omitempty"`
+}
+```
+
+### Additional required modifications
+
+Add to the VolumeSource struct:
+
+```go
+SystemProjection *SystemProjections `json:"system,omitempty"`
```

Comment:
I dislike "system" as a name - how about "projected" or "virtual" or "from" or something that is more explanatory